src/st.py

```diff
@@ -1,14 +1,134 @@
+from __future__ import annotations
 import argparse
+from dataclasses import dataclass
+import search
 
+@dataclass
+class Node:
+    start: int
+    end: int
+    child: Node | None # If end==len(x) then child is is the index
+    sib : Node | None
+
+    def __str__(self):
+        return f"{self.start},{self.end},{type(self.child)}, {type(self.sib)}"
+    
+    def __len__(self):
+        return self.end-self.start
+
+SENTINEL = "$"
+
+def getAllSuffixes(x):
+    for i in range(1, len(x)):
+        yield x[i:]+SENTINEL
+    yield SENTINEL
+
+def getMatch(x: str, y: str, n: Node, i: int):
+    nchild = n.child
+    while type(nchild) is Node:
+        if x[nchild.start] == y[i]:
+            iters = 1
+            while nchild.start + iters < nchild.end and x[nchild.start+iters] == y[i+iters]:
+                iters += 1
+            return iters, nchild
+        else:
+            nchild = nchild.sib
+    return -1, None
+
+def getDifference(str1, str2, i, j, length):
+    counter = 0
+    while counter < length and i+counter<len(str1) and j+counter<len(str2):
+        if str1[i+counter] != str2[j+counter]:
+            return counter
+        counter += 1
+    return -1
+
+def getSuffixTree(x: str):
+    suffixes = getAllSuffixes(x)
+    x += SENTINEL
```

Comment:
Why do you add the `SENTINEL` to `x` here when you also do it to every single string in `getAllSuffixes()`?